app/services/fact_cleaner.rb

```diff
@@ -0,0 +1,44 @@
+class FactCleaner
+  delegate :logger, :to => :Rails
+
+  def clean!
+    @to_delete = []
+    logger.info "Searching for orphaned facts..."
+    find_leaf_orphaned_facts
+
+    # we also delete all composes that either don't have any descendant or all of them are to be deleted
+    logger.info "Searching for compose orphaned facts..."
+    find_compose_orphaned_facts
+
+    logger.info "#{@to_delete.size} facts will be removed"
+
+    delete!
+    logger.info "Cleanup of facts finished!"
+    @to_delete.count
+  end
+
+  private
+
+  def find_leaf_orphaned_facts
+    FactName.where(:compose => false).includes(:fact_values).find_in_batches do |batch|
+      @to_delete += batch.select do |fact|
+        fact.fact_values.empty?
```

Comment:
How would you put that into scope especially until we have [rails 5 left_outer_jons](http://guides.rubyonrails.org/active_record_querying.html#left-outer-joins)? I think in this case it's not worth it, especially if we only need it for one maintenance task. I'm happy to revisit when we are on Rails 5 or there's at least one use case elsewhere to find orphaned facts.